src/codegeneration/CommonJsModuleTransformer.js

```diff
@@ -59,106 +45,103 @@ export class CommonJsModuleTransformer extends ModuleTransformer {
     return tree.moduleName;
   }
 
-  moduleProlog() {
-    let statements = super.moduleProlog();
-
-    // declare temp vars in prolog
-    if (this.moduleVars_.length) {
-      let tmpVarDeclarations = createVariableStatement(createVariableDeclarationList(VAR,
-          this.moduleVars_.map((varName) => createVariableDeclaration(varName, null))));
-
-      statements.push(tmpVarDeclarations);
+  wrapModule(statements) {
+    if (this.needsInteropRequire_) {
+      const req = parseStatement `function $__interopRequire(id) {
+        id = require(id);
+        return id && id.__esModule && id || {default: id};
+      }`;
+      return prependStatements(statements, req);
     }
-
     return statements;
   }
 
-  wrapModule(statements) {
-    let last = statements[statements.length - 1];
-    statements = statements.slice(0, -1);
-    assert(last.type === RETURN_STATEMENT);
-    let exportExpression = last.expression;
+  addExportStatement(statements) {
+    if (!this.hasExports()) {
+      return statements;
+    }
 
-    // If the module doesn't use any export statements, nor global "this", it
-    // might be because it wants to make its own changes to "exports" or
-    // "module.exports", so we don't append "module.exports = {}" to the output.
-    if (this.hasExports()) {
-      let exportStatement =
-          this.transformExportExpressionToModuleExport(exportExpression);
-      statements = statements.concat(exportStatement);
+    // The module.exports mutation happens as soon as possible,
+    // which is after all the export star are imported.
```

Comment:
I think the comment would be clearer if moved down to the `return` and strengthened:
```
// Mutate module.exports immediately after all the export star are imported,
// before any module code is executed, to allow some cyclic dependencies to work.
```